File: occasum-json/src/test/java/pw/stamina/occasum/dao/json/PropertyJsonSerializerTests.java
package pw.stamina.occasum.dao.json;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;
import name.falgout.jeffrey.testing.junit5.MockitoExtension;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import pw.stamina.occasum.properties.Property;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
final class PropertyJsonSerializerTests {

    @Mock private Property property;
    private final String valueAsString = "value";

    private PropertyJsonSerializer serializer;

    @BeforeEach
    void setupMockedPropertyAndSerializer() {
        when(property.getValueAsString()).thenReturn(valueAsString);
        serializer = new PropertyJsonSerializer();
    }

    @Test
    void serialize_givenNonNullProperty_shouldSerializePropertyGetValueAsStringToJsonPrimitive() {
        JsonElement serialized = serializer.serialize(property, null, null);
        assertTrue(serialized.isJsonPrimitive());

        JsonPrimitive primitive = serialized.getAsJsonPrimitive();
        assertEquals(valueAsString, primitive.getAsString());

        verify(property).getValueAsString();
    }
}

File: occasum-json/src/test/kotlin/pw/stamina/occasum/dao/json/PropertyJsonSerializerTests.kt
package pw.stamina.occasum.dao.json

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import pw.stamina.occasum.dao.PropertyDao
import pw.stamina.occasum.properties.Property

@ExtendWith(MockitoExtension::class)
internal class PropertyJsonSerializerTests {

    @Mock private lateinit var property: Property
    private val valueAsString = PropertyDao.RESERVED_SERIALIZED_VALUE_NAME

    private lateinit var serializer: PropertyJsonSerializer

    @BeforeEach
    fun setupMockedPropertyAndSerializer() {
        `when`(property.valueAsString).thenReturn(valueAsString)
        serializer = PropertyJsonSerializer()
    }

    @Test
    fun serialize_givenNonNullProperty_shouldSerializePropertyGetValueAsStringToJsonPrimitive() {
        val serialized = serializer.serialize(property, null, null)
        assertTrue(serialized.isJsonPrimitive)

        val primitive = serialized.asJsonPrimitive
        assertEquals(valueAsString, primitive.asString)

        verify(property).valueAsString
    }
}
